src/miniscript/satisfy.rs

```diff
@@ -212,339 +238,521 @@ impl_tuple_satisfier!(&'a A, &'b B, &'c C, &'d D, &'e E, &'f F,);
-            }
-            Terminal::AndOr(ref a, _, ref c) => {
-                let mut ret = c.node.dissatisfy(satisfier)?;
-                ret.extend(a.node.dissatisfy(satisfier)?);
-                Some(ret)
-            }
-            Terminal::OrB(ref left, ref right) | Terminal::OrD(ref left, ref right) => {
-                let mut ret = right.node.dissatisfy(satisfier)?;
-                ret.extend(left.node.dissatisfy(satisfier)?);
-                Some(ret)
-            }
-            Terminal::OrI(ref left, ref right) => {
-                match (
-                    left.node.dissatisfy(satisfier),
-                    right.node.dissatisfy(satisfier),
-                ) {
-                    (None, None) => None,
-                    (Some(mut l), None) => {
-                        l.push(vec![1]);
-                        Some(l)
-                    }
-                    (None, Some(mut r)) => {
-                        r.push(vec![1]);
-                        Some(r)
-                    }
-                    _ => panic!("tried to dissatisfy or_i but both branches were dissatisfiable"),
+            },
+            Terminal::AndB(ref l, ref r)
+                | Terminal::OrB(ref l, ref r)
+                | Terminal::OrD(ref l, ref r)
+                | Terminal::AndOr(ref l, _, ref r) => {
+                let lnsat = Self::dissatisfy(&l.node, stfr);
+                let rnsat = Self::dissatisfy(&r.node, stfr);
+                Satisfaction {
+                    stack: Witness::combine(rnsat.stack, lnsat.stack),
+                    has_sig: rnsat.has_sig || lnsat.has_sig,
                 }
-            }
+            },
+            Terminal::OrC(..) => Satisfaction {
+                stack: Witness::Unavailable,
+                has_sig: false,
+            },
+            Terminal::OrI(ref l, ref r) => {
+                let lnsat = Self::dissatisfy(&l.node, stfr);
+                let dissat_1 = Satisfaction {
+                    stack: Witness::combine(lnsat.stack, Witness::push_1()),
+                    has_sig: lnsat.has_sig,
+                };
+
+                let rnsat = Self::dissatisfy(&r.node, stfr);
+                let dissat_2 = Satisfaction {
+                    stack: Witness::combine(rnsat.stack, Witness::push_1()),
+                    has_sig: rnsat.has_sig,
+                };
+
+                Self::minimum(dissat_1, dissat_2)
+            },
             Terminal::Thresh(_, ref subs) => {
-                let mut ret = vec![];
-                for sub in subs.iter().rev() {
-                    ret.extend(sub.node.dissatisfy(satisfier)?);
+                Satisfaction {
+                    stack: subs
+                        .iter()
+                        .fold(Witness::empty(), |acc, sub| Witness::combine(
+                            Self::dissatisfy(&sub.node, stfr).stack,
+                            acc,
+                        )),
+                    has_sig: false,
```

Comment:
Should this is be any sig instead of false?

Reply:
Again, any sig should always be false (because every argument to a `Thresh` in a well-typed script is `d`). But in this case it's hard to even assert this because of the way I've structured the code -- I don't ever store the complete dissatisfactions, I just fold all the stacks into one and assume that `has_sig` is false for every one of them.

I guess I could stick an assertion into the iterator.

Reply:
@apoelstra, my logic at viewing at `thresh` is that it should share code similarity with `or_b`. Which is why I got confused by this. 
I guess it makes sense to club the `or_b` with `and_b` since has_sig for `or_b` dissat should be false but is written explicitly. 
Anyway, it does not matter for correctness of the code. So, everything should be fine.

Reply:
I added an assert on every individual (dis)satisfaction to check that it has no signatures.